igvm/hypervisor_preferences.py

```diff
@@ -219,16 +278,104 @@ def __call__(self, vm, hv):
         tgt_hv_rl_cpus = hv.dataset_obj[self.attribute]
         tgt_ovr_allc = float(tgt_hv_cpus) / float(tgt_hv_rl_cpus)
 
-        return tgt_ovr_allc > cur_ovr_allc
+        # whether the target hv would be more overbooked than the current one
+        rel_overbooking = tgt_ovr_allc / cur_ovr_allc
+        if rel_overbooking > 1.:
+            return .01
+
+        # normalize the value. this is no hard criteria, except for non-
+        # overbookable resources like disk. hence we can allow overbooked cpus.
+        # for cpus the avg load is more important, for anything else non-
+        # overbookable we will leave the corresponding hard-check for later
+        return 1 - rel_overbooking
+
+
+class PreferenceEvaluator(object):
+    def __init__(self, vm, hv, preferences: List[HypervisorPreference]):
+        self.vm = vm
+        self.hv = hv
+        self.preferences = preferences
+        self.total: Optional[float] = None
+
+    def get_preference(self) -> float:
+        if self.total:
+            return self.total
+
+        n_prefs = len(self.preferences)
+        matched_prefs = 0
+        sum_prefs = 0.
+
+        for p in self.preferences:
+            result = float(p.get_preference(self.vm, self.hv))
```

Comment:
Methods would fail before if they do not return a float because of the type hints.

Reply:
type hints do not make your code fail. they are just useful for static code analysis and documentation purposes